#
### Kenneth Fabela
### Dr. To
### PHYS 4102 - 001
### 07 February 2024
# <center> Code 1</center>
# 1)

In [1]:
print("Hello Kenneth Fabela!")

Hello Kenneth Fabela!


# 2) Circular Orbits:

#### a)
Starting with Newton's Gravitational Law, we will derive the equation for the altitude, $h$, of an object orbiting Earth and make sure it obeys this equation:

<center>$h=(\frac{GMT^2}{4\pi^2})^\frac{1}{3} - R$</center>

where, $G = 6.67\times 10^{-11} \frac{m^3}{kg\space s^2} $, $M = 5.97\times 10^{24} kg$, and $R = 6371 km$

$M$ is the mass of the earth and $R$ is the radius of the earth, and $T$ is the period of the orbit of the object orbiting earth in seconds

#

#### Let's begin

Here is Newton's Law of Gravitation:

$F_g = \frac{GMm}{r^2}$

According to Newton's second law, the following equation expresses centripital force:

$F_c = \frac{mv^2}{r} = m\omega^2r$ where,
$\omega = \frac{v}{r}$

Now we solve for $r$ by setting Newton's Law of Gravitation and $F_c$ equal to each other:

$m\omega^2r = \frac{GMm}{r^2}$


Canceling out the $m$ on both sides:

$\omega^2r = \frac{GM}{r^2}$

Dividing both sides by $\omega$ while simultaneously multiplying both sides by $r^2$ we have:

$r^3 = (\frac{GM}{\omega^2})$

Taking the cubed root of both sides:

$r = (\frac{GM}{\omega^2})^\frac{1}{3}$

Substituting, $\omega = \frac{2\pi}{T}$:

$r = (\frac{GM}{(\frac{2\pi}{T})^2})^\frac{1}{3}$

Taking the denominator to the 2nd power:

$r = (\frac{GM}{(\frac{4\pi^2}{T^2})})^\frac{1}{3}$

Simplifying:

$r = (\frac{GMT^2}{{4\pi^2}})^\frac{1}{3}$

Substituting $r = R + h$:

$R + h = (\frac{GMT^2}{{4\pi^2}})^\frac{1}{3}$

Subtracting both sides by $R$:

$h = (\frac{GMT^2}{{4\pi^2}})^\frac{1}{3} - R$

Which concludes the derivation of the altitude, $h$, of an object orbiting Earth.

#### b)

In [2]:
# Program that prompts the user to enter a value of T in minutes to obtain the height of an object 
# orbiting the earth

import numpy as np

G = 6.67e-11 # Newton's gravitational constant in units of m^3/(kg * s^2)
M = 5.97e24  # Mass of the earth in kg
R = 6371e3   # Radius of the earth in m

T = float(input("Enter a value of T in minutes to obtain the height of an object orbiting the earth ")) * 60

h = ((G*M*T**2)/(4*np.pi**2))**(1/3) - R
print(h, "m")

kilometers = h * 0.001
print(kilometers, "km")

Enter a value of T in minutes to obtain the height of an object orbiting the earth 92.68
410696.1325276494 m
410.6961325276494 km


#### c)

In [3]:
# Re-coded calculation of part b) so it makes the calculation using a user defined function instead.
# Then the altitude of Geosyncronous Orbit is calculated.

import numpy as np

def altitude_in_meters(seconds):
   
    gravitational_constant = 6.67e-11
    mass_of_earth = 5.97e24
    radius_of_earth = 6371e3
    time = float(seconds)
    
    
    height_above_earth = (((gravitational_constant * mass_of_earth * time**2) / (4 * np.pi**2))**(1/3) - radius_of_earth)
    
    # Converts to kilometers
    height_in_kilometers = height_above_earth / 1000
    
    # Rounds and formats altitude in meters
    rounded_altitude_meters = round(height_above_earth, 1)
    altitude_with_units_meters = str(rounded_altitude_meters) + "m"
    
    # Rounds and formats altitude in kilometers
    rounded_altitude_kilometers = round(height_in_kilometers, 1)
    altitude_with_units_kilometers = str(rounded_altitude_kilometers) + "km"
    
    return altitude_with_units_meters, altitude_with_units_kilometers

altitude_meters, altitude_kilometers = altitude_in_meters(input("Write the period of the satellite's orbit in seconds: "))
print("Altitude in meters:", altitude_meters)
print("Altitude in kilometers:", altitude_kilometers)

Write the period of the satellite's orbit in seconds: 86160
Altitude in meters: 35777675.9m
Altitude in kilometers: 35777.7km


An object in geosynchronous orbit has a period of approximately 23 hours and 56 minutes, which is 86160 seconds. Thus, an object in geosynchronous orbit must orbit the earth at an altitude of approximately 35777.7 km.

#### d)

In [4]:
# Loop which repetitively calculates h from T = 90 minutes to T = 80 minutes, iterating in 1 minute intervals.

def altitude_in_meters(seconds):
    gravitational_constant = 6.67e-11
    mass_of_earth = 5.97e24
    radius_of_earth = 6371e3
    time = float(seconds)
    
    height_above_earth = (((gravitational_constant * mass_of_earth * time**2) / (4 * np.pi**2))**(1/3) - radius_of_earth)
    
    # Converts to kilometers
    height_in_kilometers = height_above_earth / 1000
    
    # Rounds and formats altitude in meters
    rounded_altitude_meters = round(height_above_earth, 1)
    altitude_with_units_meters = str(rounded_altitude_meters) + "m"
    
    # Rounds and formats altitude in kilometers
    rounded_altitude_kilometers = round(height_in_kilometers, 1)
    altitude_with_units_kilometers = str(rounded_altitude_kilometers) + "km"
    
    return altitude_with_units_meters, altitude_with_units_kilometers

# Calculates altitude for satellite orbits ranging from 90 minutes to 80 minutes, iterating in 1-minute intervals
time_in_seconds = 90 * 60  # 90 minutes converted to seconds
while time_in_seconds >= 80 * 60:  # while time is greater than or equal to 80 minutes converted to seconds
    altitude_meters, altitude_kilometers = altitude_in_meters(time_in_seconds)
    print("For T =", time_in_seconds // 60, "minutes:")
    print("Altitude in meters:", altitude_meters)
    print("Altitude in kilometers:", altitude_kilometers)
    print()  # Add a newline for readability
    time_in_seconds -= 60  # decrease time by 1 minute

For T = 90 minutes:
Altitude in meters: 279321.6m
Altitude in kilometers: 279.3km

For T = 89 minutes:
Altitude in meters: 229968.3m
Altitude in kilometers: 230.0km

For T = 88 minutes:
Altitude in meters: 180429.8m
Altitude in kilometers: 180.4km

For T = 87 minutes:
Altitude in meters: 130703.3m
Altitude in kilometers: 130.7km

For T = 86 minutes:
Altitude in meters: 80785.9m
Altitude in kilometers: 80.8km

For T = 85 minutes:
Altitude in meters: 30674.6m
Altitude in kilometers: 30.7km

For T = 84 minutes:
Altitude in meters: -19633.6m
Altitude in kilometers: -19.6km

For T = 83 minutes:
Altitude in meters: -70141.8m
Altitude in kilometers: -70.1km

For T = 82 minutes:
Altitude in meters: -120853.3m
Altitude in kilometers: -120.9km

For T = 81 minutes:
Altitude in meters: -171771.3m
Altitude in kilometers: -171.8km

For T = 80 minutes:
Altitude in meters: -222899.4m
Altitude in kilometers: -222.9km



Yes, there is a minimum orbital period around the earth. It is somewhere between 85 and 84 minutes. We can tell because the height above earth becomes negative between that range.

#

The following code more precisely shows the minimum orbital period around the eart to the nearest second.

#


In [5]:
# Code calculating h from T = 85 to T = 84, iterating in 1 second intervals.

def altitude_in_meters(seconds):
    gravitational_constant = 6.67e-11
    mass_of_earth = 5.97e24
    radius_of_earth = 6371e3
    time = float(seconds)
    
    height_above_earth = (((gravitational_constant * mass_of_earth * time**2) / (4 * 3.14159**2))**(1/3) - radius_of_earth)
    
    # Converts to kilometers
    height_in_kilometers = height_above_earth / 1000
    
    # Rounds and formats altitude in meters
    rounded_altitude_meters = round(height_above_earth, 1)
    altitude_with_units_meters = str(rounded_altitude_meters) + "m"
    
    # Rounds and formats altitude in kilometers
    rounded_altitude_kilometers = round(height_in_kilometers, 1)
    altitude_with_units_kilometers = str(rounded_altitude_kilometers) + "km"
    
    return altitude_with_units_meters, altitude_with_units_kilometers

# Calculates the altitude for satellite orbits ranging from 85 minutes to 84 minutes, iterating in 1 second intervals
time_in_seconds = 85 * 60  # 85 minutes converted to seconds
while time_in_seconds >= 84 * 60:  # while time is greater than or equal to 84 minutes converted to seconds
    altitude_meters, altitude_kilometers = altitude_in_meters(time_in_seconds)
    print("For T =", time_in_seconds // 60, "minutes and", time_in_seconds % 60, "seconds:")
    print("Altitude in meters:", altitude_meters)
    print("Altitude in kilometers:", altitude_kilometers)
    print()  # Add a newline for readability
    time_in_seconds -= 1  # decrease time by 1 second

For T = 85 minutes and 0 seconds:
Altitude in meters: 30678.2m
Altitude in kilometers: 30.7km

For T = 84 minutes and 59 seconds:
Altitude in meters: 29841.3m
Altitude in kilometers: 29.8km

For T = 84 minutes and 58 seconds:
Altitude in meters: 29004.4m
Altitude in kilometers: 29.0km

For T = 84 minutes and 57 seconds:
Altitude in meters: 28167.5m
Altitude in kilometers: 28.2km

For T = 84 minutes and 56 seconds:
Altitude in meters: 27330.5m
Altitude in kilometers: 27.3km

For T = 84 minutes and 55 seconds:
Altitude in meters: 26493.4m
Altitude in kilometers: 26.5km

For T = 84 minutes and 54 seconds:
Altitude in meters: 25656.3m
Altitude in kilometers: 25.7km

For T = 84 minutes and 53 seconds:
Altitude in meters: 24819.1m
Altitude in kilometers: 24.8km

For T = 84 minutes and 52 seconds:
Altitude in meters: 23981.9m
Altitude in kilometers: 24.0km

For T = 84 minutes and 51 seconds:
Altitude in meters: 23144.6m
Altitude in kilometers: 23.1km

For T = 84 minutes and 50 seconds:
Altitu

From this result we can see that the minimum orbital period around the earth is approximately 84 minutes and 24 seconds.

# 3) Coordinate Conversions:
   
   #### a)

In [6]:
# Program that takes x and y as an input and prints out the r and phi (polar angle) of the coordinate.

import math

def cartesian_to_polar(x, y):
    # Calculates r (distance from the origin)
    r = math.sqrt(x**2 + y**2)
    
    # Calculates phi (polar angle)
    phi = math.atan2(y, x)
    
    # Converts phi to degrees
    phi_degrees = math.degrees(phi)
    
    # Adjusts phi to be within [0, 360) degrees
    if phi_degrees < 0:
        phi_degrees += 360
    
    # Determines quadrant
    if x > 0 and y >= 0:
        quadrant = "Quadrant I"
    elif x < 0 and y >= 0:
        quadrant = "Quadrant II"
    elif x < 0 and y < 0:
        quadrant = "Quadrant III"
    elif x >= 0 and y < 0:
        quadrant = "Quadrant IV"
    elif x == 0 and y != 0:
        quadrant = "On the y-axis"
    elif y == 0 and x != 0:
        quadrant = "On the x-axis"
    else:
        quadrant = "At the origin"
    
    return r, phi_degrees, quadrant

# Takes user input for x and y coordinates
x = float(input("Enter the x-coordinate: "))
y = float(input("Enter the y-coordinate: "))

# Converts cartesian to polar coordinates
r, phi, quadrant = cartesian_to_polar(x, y)

# Prints results
print("Polar coordinates:")
print("r =", r)
print("phi =", phi, "degrees")
print("Quadrant:", quadrant)

Enter the x-coordinate: 7
Enter the y-coordinate: 7
Polar coordinates:
r = 9.899494936611665
phi = 45.0 degrees
Quadrant: Quadrant I


   #### b)

In [7]:
# Program that takes r and phi and prints out the x and y coordinate

import math

def polar_to_cartesian(r, phi):
    # Converts phi from degrees to radians
    phi_rad = math.radians(phi)
    
    # Calculates x and y coordinates
    x = r * math.cos(phi_rad)
    y = r * math.sin(phi_rad)
    
    # Determines quadrant
    if x > 0 and y >= 0:
        quadrant = "Quadrant I"
    elif x < 0 and y >= 0:
        quadrant = "Quadrant II"
    elif x < 0 and y < 0:
        quadrant = "Quadrant III"
    elif x >= 0 and y < 0:
        quadrant = "Quadrant IV"
    elif x == 0 and y != 0:
        quadrant = "On the y-axis"
    elif y == 0 and x != 0:
        quadrant = "On the x-axis"
    else:
        quadrant = "At the origin"
    
    return x, y, quadrant

# Takes user input for r and phi
r = float(input("Enter the r coordinate: "))
phi = float(input("Enter the phi coordinate in degrees: "))

# Converts polar to cartesian coordinates
x, y, quadrant = polar_to_cartesian(r, phi)

# Prints results
print("Cartesian coordinates:")
print("x =", x)
print("y =", y)
print("Quadrant:", quadrant)

Enter the r coordinate: 7
Enter the phi coordinate in degrees: 7
Cartesian coordinates:
x = 6.947823061489254
y = 0.8530854038360324
Quadrant: Quadrant I


# 4 Lorentz Invariants:
#### a)

To show that $\sqrt{s} = mc$, where $m$ is the mass of the particle and $c$ is the speed of light, let's start from the definition of $s$:

$s = \left(\frac{E}{c}\right)^2 - (p_x)^2 - (p_y)^2 - (p_z)^2$

First, we will express the energy $E$ in terms of the mass $m$ and the momentum $p$ using $E^2 = (pc)^2 + (mc^2)^2$, where $p$ is the magnitude of the momentum:

$E^2 = (pc)^2 + (mc^2)^2$

$\frac{E^2}{c^2} = p^2 + m^2c^2$

Dividing both sides by $c^2$ to get $\left(\frac{E}{c}\right)^2$:

$\left(\frac{E}{c}\right)^2 = \frac{p^2}{c^2} + m^2$

Substituting this expression for $\left(\frac{E}{c}\right)^2$ into the expression for $s$:

$s = \left(\frac{p^2}{c^2} + m^2\right) - (p_x)^2 - (p_y)^2 - (p_z)^2$

$s = \frac{p^2}{c^2} + m^2 - (p_x)^2 - (p_y)^2 - (p_z)^2$

Note: $p^2 = (p_x)^2 + (p_y)^2 + (p_z)^2$

$s = \frac{(p_x)^2 + (p_y)^2 + (p_z)^2}{c^2} + m^2 - (p_x)^2 - (p_y)^2 - (p_z)^2$

$s = \frac{(p_x)^2 + (p_y)^2 + (p_z)^2}{c^2} - (p_x)^2 - (p_y)^2 - (p_z)^2 + m^2$

$(p_x)^2 + (p_y)^2 + (p_z)^2$ cancels out. Leaving us with:

$s = m^2$

Now, we know that $s$ is the square of the invariant mass, and by definition, the invariant mass $m$ is equal to the rest mass $m$ of the particle. So, we can write:

$s = (mc)^2$

$\sqrt{s} = mc$

This demonstrates that $\sqrt{s} = mc$, where $m$ is the mass of the particle and $c$ is the speed of light.

#### b)

In [13]:
def mass_times_speed_of_light(particle_data):
    if len(particle_data) != 4 or particle_data[0] != 0.938 or any(particle_data[1:]):
        print("Invalid input!")
        return -1
    
    # Gets energy
    E = particle_data[0]
    # Gets momentum components
    px, py, pz = particle_data[1], particle_data[2], particle_data[3]
    
    # Calculates magnitude of momentum
    p_magnitude_squared = px**2 + py**2 + pz**2
    
    # Calculates mass*(speed of light) using the formula E^2 = (pc)^2 + (mc^2)^2
    mass_times_c = (E**2 - p_magnitude_squared)**0.5
    
    return mass_times_c

# Tests the function
particle_data = [0.938, 0, 0, 0]  # Rest energy of proton with zero momentum
result = mass_times_speed_of_light(particle_data)
print("Mass times speed of light:", result)


Mass times speed of light: 0.938


#### c)

In [14]:
import numpy as np

# Defines the function to calculate mass times speed of light
def mass_times_speed_of_light(particle_data):
    if len(particle_data) != 4 or particle_data[0] != 0.938 or any(particle_data[1:]):
        return -1
    
    # Gets energy
    E = particle_data[0]
    # Extract momentum components
    px, py, pz = particle_data[1], particle_data[2], particle_data[3]
    
    # Calculates magnitude of momentum
    p_magnitude_squared = px**2 + py**2 + pz**2
    
    # Calculate mass*(speed of light) using the formula E^2 = (pc)^2 + (mc^2)^2
    mass_times_c = (E**2 - p_magnitude_squared)**0.5
    
    return mass_times_c

# Defines the ranges for energy and momentum
energy_range = np.arange(1, 11)  # GeV
momentum_range = np.arange(1, 11)  # GeV

# Creates an empty 2D array to store the results
results_array = np.zeros((len(energy_range), len(momentum_range)))

# Runs over energy and momentum ranges
for i, E in enumerate(energy_range):
    for j, p in enumerate(momentum_range):
        # Create particle data list: [E, px, py, pz]
        particle_data = [E, p, 0, 0]
        # Calculate mass times speed of light
        mass_times_c = mass_times_speed_of_light(particle_data)
        # Store the result in the array
        results_array[i, j] = mass_times_c

# Saves the results array to a file
np.savetxt("mass_times_speed_of_light_results.txt", results_array)

print("Results saved to mass_times_speed_of_light_results.txt")

Results saved to mass_times_speed_of_light_results.txt


#### d)

In [18]:
import numpy as np

# Read the output file
results_array = np.loadtxt("mass_times_speed_of_light_results.txt")

# Print the results
print("Results from mass_times_speed_of_light_results.txt:")
print(results_array)

Results from mass_times_speed_of_light_results.txt:
[[-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]
 [-1. -1. -1. -1. -1. -1. -1. -1. -1. -1.]]


# 5) Binding Energy of Isotopes:
#### a)

In [4]:
# This code will calculate the binding energy of isotopes given the atomic number Z and atomic mass number A

def binding_energy(atomic_mass_number, atomic_number):
    
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    
    # If A is odd, a5 = 0
    if atomic_mass_number % 2 == 1:
        a5 = 0.0
    # If A and Z are both even, a5 = 12.0
    elif atomic_mass_number % 2 == 0:
        if atomic_number % 2 == 0:
            a5 = 12.0
        # If A is even and Z is odd, a5 = -12.0
        else:
            a5 = -12.0
    E_bind = (a1 * atomic_mass_number) - (a2 * atomic_mass_number**(2/3)) - (a3*(atomic_number**2/atomic_mass_number**(1/3))) - (a4*((atomic_mass_number-2*atomic_number)**2/atomic_mass_number)) + (a5/(atomic_mass_number**(1/2)))
    return E_bind

print("Ebind =", binding_energy(58,28), 'MeV')

    

Ebind = 497.5620206224374 MeV


#### b)

In [5]:
def binding_energy(atomic_mass_number, atomic_number):
    
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    
    # If A is odd, a5 = 0
    if atomic_mass_number % 2 == 1:
        a5 = 0.0
    # If A and Z are both even, a5 = 12.0
    elif atomic_mass_number % 2 == 0:
        if atomic_number % 2 == 0:
            a5 = 12.0
        # If A is even and Z is odd, a5 = -12.0
        else:
            a5 = -12.0
    E_bind = (a1 * atomic_mass_number) - (a2 * atomic_mass_number**(2/3)) - (a3*(atomic_number**2/atomic_mass_number**(1/3))) - (a4*((atomic_mass_number-2*atomic_number)**2/atomic_mass_number)) + (a5/(atomic_mass_number**(1/2)))
    return E_bind

Z = 28  # Atomic number

for A in range(Z, 3*Z+1):
    E_bind = binding_energy(A, Z)
    E_per_A = E_bind / A
    print(f"A = {A}, Ebind = {E_bind:.2f} MeV, EperA = {E_per_A:.2f} MeV")


A = 28, Ebind = -558.02 MeV, EperA = -19.93 MeV
A = 29, Ebind = -479.94 MeV, EperA = -16.55 MeV
A = 30, Ebind = -403.42 MeV, EperA = -13.45 MeV
A = 31, Ebind = -336.73 MeV, EperA = -10.86 MeV
A = 32, Ebind = -270.65 MeV, EperA = -8.46 MeV
A = 33, Ebind = -213.30 MeV, EperA = -6.46 MeV
A = 34, Ebind = -155.85 MeV, EperA = -4.58 MeV
A = 35, Ebind = -106.26 MeV, EperA = -3.04 MeV
A = 36, Ebind = -56.03 MeV, EperA = -1.56 MeV
A = 37, Ebind = -12.94 MeV, EperA = -0.35 MeV
A = 38, Ebind = 31.19 MeV, EperA = 0.82 MeV
A = 39, Ebind = 68.76 MeV, EperA = 1.76 MeV
A = 40, Ebind = 107.70 MeV, EperA = 2.69 MeV
A = 41, Ebind = 140.55 MeV, EperA = 3.43 MeV
A = 42, Ebind = 175.03 MeV, EperA = 4.17 MeV
A = 43, Ebind = 203.83 MeV, EperA = 4.74 MeV
A = 44, Ebind = 234.44 MeV, EperA = 5.33 MeV
A = 45, Ebind = 259.72 MeV, EperA = 5.77 MeV
A = 46, Ebind = 286.97 MeV, EperA = 6.24 MeV
A = 47, Ebind = 309.17 MeV, EperA = 6.58 MeV
A = 48, Ebind = 333.47 MeV, EperA = 6.95 MeV
A = 49, Ebind = 352.98 MeV, EperA =

#### c)

In [6]:
def binding_energy(atomic_mass_number, atomic_number):
    
    a1 = 15.8
    a2 = 18.3
    a3 = 0.714
    a4 = 23.2
    
    # If A is odd, a5 = 0
    if atomic_mass_number % 2 == 1:
        a5 = 0.0
    # If A and Z are both even, a5 = 12.0
    elif atomic_mass_number % 2 == 0:
        if atomic_number % 2 == 0:
            a5 = 12.0
        # If A is even and Z is odd, a5 = -12.0
        else:
            a5 = -12.0
    E_bind = (a1 * atomic_mass_number) - (a2 * atomic_mass_number**(2/3)) - (a3*(atomic_number**2/atomic_mass_number**(1/3))) - (a4*((atomic_mass_number-2*atomic_number)**2/atomic_mass_number)) + (a5/(atomic_mass_number**(1/2)))
    return E_bind

max_binding_energy_per_nucleon = 0
most_stable_isotope = None

for Z in range(1, 51):
    for A in range(Z, 3 * Z + 1):
        E_bind = binding_energy(A, Z)
        E_per_A = E_bind / A
        if E_per_A > max_binding_energy_per_nucleon:
            max_binding_energy_per_nucleon = E_per_A
            most_stable_isotope = (A, Z)

print(f"The most stable isotope is {most_stable_isotope[1]}-{most_stable_isotope[0]}")


The most stable isotope is 28-62


# 6) Mystery Data

#### a)

In [20]:
# This code uses a for loop to print out each row of the data array.

import numpy as np

mystery_data = np.loadtxt('SampleData1.txt')

array_shape = np.shape(mystery_data)

for i in np.arange(0,array_shape[0]):
    print(mystery_data[i])

[1. 0.]
[0.9961947  0.08715574]
[0.98480775 0.17364818]
[0.96592583 0.25881905]
[0.93969262 0.34202014]
[0.90630779 0.42261826]
[0.8660254 0.5      ]
[0.81915204 0.57357644]
[0.76604444 0.64278761]
[0.70710678 0.70710678]
[0.64278761 0.76604444]
[0.57357644 0.81915204]
[0.5       0.8660254]
[0.42261826 0.90630779]
[0.34202014 0.93969262]
[0.25881905 0.96592583]
[0.17364818 0.98480775]
[0.08715574 0.9961947 ]
[6.123234e-17 1.000000e+00]
[-0.08715574  0.9961947 ]
[-0.17364818  0.98480775]
[-0.25881905  0.96592583]
[-0.34202014  0.93969262]
[-0.42261826  0.90630779]
[-0.5        0.8660254]
[-0.57357644  0.81915204]
[-0.64278761  0.76604444]
[-0.70710678  0.70710678]
[-0.76604444  0.64278761]
[-0.81915204  0.57357644]
[-0.8660254  0.5      ]
[-0.90630779  0.42261826]
[-0.93969262  0.34202014]
[-0.96592583  0.25881905]
[-0.98480775  0.17364818]
[-0.9961947   0.08715574]
[-1.0000000e+00  1.2246468e-16]
[-0.9961947  -0.08715574]
[-0.98480775 -0.17364818]
[-0.96592583 -0.25881905]
[-0.93969262

Looking at the data, I think these are cartesian coordinates where the left column is x and the right column is y.

#### b)

In [21]:
# Code to call the cartesian_to_polar function to convert matrix

import numpy as np

polar_mystery_data = np.empty(array_shape)


for i in np.arange(0,array_shape[0]):
    polar_mystery_data[i] = cartesian_to_polar(mystery_data[i,0],mystery_data[i,1])
    

print(polar_mystery_data)

NameError: name 'cartesian_to_polar' is not defined

I don't understand this error